language/bytecode_verifier/src/instantiation_loops.rs

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This implements an algorithm that detects loops during the instantiation of generics.
+//!
+//! It builds a graph from the given `CompiledModule` and converts the original problem into
+//! finding strongly connected components in the graph with certain properties. Read the
+//! documentation of the types/functions below for details of how it works.
+
+use petgraph::{
+    algo::tarjan_scc,
+    graph::{EdgeIndex, NodeIndex},
+    visit::EdgeRef,
+    Graph,
+};
+use std::collections::{hash_map, HashMap, HashSet};
+use types::vm_error::{StatusCode, VMStatus};
+use vm::{
+    access::ModuleAccess,
+    file_format::{
+        Bytecode, CompiledModule, FunctionHandleIndex, ModuleHandleIndex, SignatureToken,
+        TypeParameterIndex,
+    },
+};
+
+/// Data attached to each node.
+/// Each node corresponds to a type formal of a generic function in the module.
+#[derive(Eq, PartialEq, Hash, Copy, Clone)]
+struct Node(FunctionHandleIndex, TypeParameterIndex);
+
+/// Data attached to each edge. Indicating the type of the edge.
+enum Edge<'a> {
+    /// This type of edge from type formal T1 to T2 means the type bound to T1 is used to
+    /// instantiate T2 unmodified, thus the name `Identity`.
+    ///
+    /// Example:
+    /// //    foo<T>() { bar<T>(); return; }
+    /// //
+    /// //    edge: foo_T -> Id -> bar_T
+    Identity,
+    /// This type of edge from type formal T1 to T2 means T2 is instantiated with a type resulted
+    /// by applying one or more type constructors to T1 (potentially with other types).
+    ///
+    /// This is interesting to us as it creates a new (and bigger) type.
+    ///
+    /// Example:
+    /// //    struct Baz<T> {}
+    /// //    foo<T>() { bar<Baz<T>>(); return; }
+    /// //
+    /// //    edge: foo_T -> TyConApp -> bar_T
+    TyConApp(&'a SignatureToken),
+}
+
+pub struct InstantiationLoopChecker<'a> {
+    module: &'a CompiledModule,
+
+    graph: Graph<Node, Edge<'a>>,
+    node_map: HashMap<Node, NodeIndex>,
+}
+
+impl<'a> InstantiationLoopChecker<'a> {
+    pub fn new(module: &'a CompiledModule) -> Self {
+        Self {
+            module,
+            graph: Graph::new(),
+            node_map: HashMap::new(),
+        }
+    }
+
+    /// Retrives the node corresponding to the specified type formal.
+    /// If none exists in the graph yet, create one.
+    fn get_or_add_node(&mut self, node: Node) -> NodeIndex {
+        match self.node_map.entry(node) {
+            hash_map::Entry::Occupied(entry) => *entry.get(),
+            hash_map::Entry::Vacant(entry) => {
+                let idx = self.graph.add_node(node);
+                entry.insert(idx);
+                idx
+            }
+        }
+    }
+
+    /// Helper function that extracts type parameters from a given type.
+    /// Duplicated entries are removed.
+    fn extract_type_parameters(ty: &SignatureToken) -> Vec<TypeParameterIndex> {
```

Comment:
Could you use a `BTreeSet` here instead? Could prevent the `sort()` and `dedup()` calls?

Reply:
Talked offline. Will explore HashSet.